### PR TITLE
Add fp4 support(Currently static double quantization is not working)

### PR DESCRIPTION
**Pull request by lanluo-nvidia**
# Description

Add fp4 support

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 1]
```diff
@@ -250,13 +250,10 @@ def cross_compile_for_windows(
...
-    if use_explicit_typing:
-        if len(enabled_precisions) != 1 or not any(
-            x in enabled_precisions for x in {torch.float32, dtype.f32}
-        ):
-            raise AssertionError(
-                f"When use_explicit_typing is enabled, only torch.float32 is allowed in the enabled_precisions but found {enabled_precisions}"
-            )
+    if use_explicit_typing and len(enabled_precisions) != 1:
+        raise AssertionError(
+            f"When use_explicit_typing is enabled, allow only 1 precision in the enabled_precisions but found {enabled_precisions}"
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py]: What does this error mean?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 2]
```diff
@@ -581,13 +578,10 @@ def compile(
...
 
-    if use_explicit_typing:
-        if len(enabled_precisions) != 1 or not any(
-            x in enabled_precisions for x in {torch.float32, dtype.f32}
-        ):
-            raise AssertionError(
-                f"When use_explicit_typing is enabled, only torch.float32 is allowed in the enabled_precisions but found {enabled_precisions}"
-            )
+    if use_explicit_typing and len(enabled_precisions) != 1:
+        raise AssertionError(
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py]: Same here

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py [hunk 3]
```diff
@@ -39,7 +39,7 @@
     CallingConvention,
 )
 from torch_tensorrt.dynamo.conversion._TRTBuilderMonitor import TRTBulderMonitor
-from torch_tensorrt.dynamo.conversion.converter_utils import (
+from torch_tensorrt.dynamo.conversion.converter_utils import (  # global_reference_holder,
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py]: What is this comment?

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py [hunk 4]
```diff
@@ -274,16 +274,28 @@ def _populate_trt_builder_config(
                 self.compilation_settings.dla_global_dram_size,
             )
 
-        if dtype.float16 in self.compilation_settings.enabled_precisions:
+        if (
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py]: Do you have the expected semantics of these models written out somewhere?

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py [hunk 5]
```diff
@@ -737,7 +756,8 @@ def run(
             self.ctx.net, builder_config
         )
         assert serialized_engine
-
+        # clear the global reference holder after engin is built
+        # global_reference_holder.clear()
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py]: What are these comments?

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 6]
```diff
@@ -617,6 +617,42 @@ def aten_ops_quantize_op(
         )
 
 
+try:
+    import modelopt.torch.quantization as mtq  # noqa: F401
+
+    assert torch.ops.tensorrt.dynamic_block_quantize_op.default
+except Exception as e:
+    _LOGGER.warning(
+        "Unable to import dynamic block quantize op. Please install modelopt library (https://github.com/NVIDIA/TensorRT-Model-Optimizer?tab=readme-ov-file#installation) to add support for compiling dynamic blockquantized models"
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py]: Are dynamic block quantized models something that people will know?

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 7]
```diff
@@ -617,6 +617,42 @@ def aten_ops_quantize_op(
...
+    )
+else:
+
+    @dynamo_tensorrt_converter(
+        torch.ops.tensorrt.dynamic_block_quantize_op.default,
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:633]: Is the only use of this op for fp4 or will it be used for other types in the future?

File: py/torch_tensorrt/dynamo/conversion/converter_utils.py [hunk 8]
```diff
@@ -25,6 +25,9 @@
 
 _LOGGER: logging.Logger = logging.getLogger(__name__)
 
+# global reference holder for the objects that should not be garbage collected before the engine is built
+global_reference_holder: List[torch.Tensor] = []
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/converter_utils.py]: You should make this a member of the conversion context not just some global buffer. When the conversion context is cleaned up it will release these

Comment (narendasan) [reply py/torch_tensorrt/dynamo/conversion/converter_utils.py]: Really we should be explicitly cleaning these so maybe as a post compilation clean up step

File: py/torch_tensorrt/dynamo/conversion/converter_utils.py [hunk 9]
```diff
@@ -361,12 +364,37 @@ def create_constant(
             shape = list(torch_value.shape)
 
         if torch_value is not None:
+            if torch_value.dtype in (torch.float8_e4m3fn, torch.uint8):
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/converter_utils.py]: Why do we handle both fp8 and fp4 in this one conditional instead of separate ones?

File: py/torch_tensorrt/dynamo/conversion/converter_utils.py [hunk 10]
```diff
@@ -361,12 +364,37 @@ def create_constant(
...
+                    ptr=torch_value.data_ptr(),
+                    count=count,
+                )
+                constant = ctx.net.add_constant(
+                    shape,
+                    weights,
+                )
+                constant.name = name
+                # TODO: confirm with @dheeraj @naren whether i can use ctx.mapping as the reference holder to prevent the torch tensor being garbage collected.
+                ctx.mapping[name + " CONSTANT"] = torch_value.reshape(-1)
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/converter_utils.py]: This isnt really the point of this dictionary. It is used to create a mapping for weight refitting. It might be a side effect that it serves as a reference holder. @cehongwang Is this the correct usage for weight refitting?

Comment (narendasan) [reply py/torch_tensorrt/dynamo/conversion/converter_utils.py]: We can either create two dictionaries because supporting weight refitting is probably good and we can have another which actually does the reference holding. so @cehongwang can provide guidance on correct usage of this map. 

also @cehongwang please rename this to something that is more descriptive like, `weight_src_to_constant_map` or something so people dont get confused

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py (hunk 6 above)
Comment (lanluo-nvidia) [reply py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py]: modified to be more generic:
`"Unable to import quantize op. Please install modelopt library (https://github.com/NVIDIA/TensorRT-Model-Optimizer?tab=readme-ov-file#installation) to add support for compiling quantized models"`

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py (hunk 7 above)
Comment (lanluo-nvidia) [reply py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:633]: currently it is only for fp4, but in future it can add more dynamic block quantization types.
changed file name from nvfp4_quantize.py to dynamic_block_quantize.py

File: py/torch_tensorrt/dynamo/conversion/converter_utils.py (hunk 8 above)
Comment (lanluo-nvidia) [reply py/torch_tensorrt/dynamo/conversion/converter_utils.py]: removed from here, added weights_reference_holder in ctx: ConversionContext

File: py/torch_tensorrt/dynamo/conversion/converter_utils.py (hunk 9 above)
Comment (lanluo-nvidia) [reply py/torch_tensorrt/dynamo/conversion/converter_utils.py]: split into separated if

File: py/torch_tensorrt/dynamo/conversion/converter_utils.py (hunk 10 above)
Comment (lanluo-nvidia) [reply py/torch_tensorrt/dynamo/conversion/converter_utils.py]: removed from here, added weights_reference_holder in ctx: ConversionContext

File: py/torch_tensorrt/dynamo/conversion/impl/dynamic_block_quantize.py [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+    on the output_type set and dequantizes them back.
+    """
+    if len(input_tensor.shape) not in (2, 3):
+        raise ValueError(
+            f"dynamic_block_quantize converter received an input of {input_tensor.shape} shape. Supported shapes: 2D or 3D"
+        )
+    with unset_fake_temporarily():
+        axis = -1
+        global_scale = _calculate_global_scale(ctx, name, amax)
+        if ".weight_quantizer" in name:
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/impl/dynamic_block_quantize.py:41]: Is this really the way we determine which quantization scheme to use? What metadata is associated with these nodes?

File: tests/py/dynamo/models/test_models_export.py [hunk 12]
```diff
@@ -199,6 +199,133 @@ def test_resnet18_half(ir):
     torch._dynamo.reset()
 
 
+# @unittest.skipIf(
+#     torch.cuda.get_device_capability() < (10, 0),
+#     "FP4 quantization requires compute capability 10.0 or later",
+# )
+@unittest.skipIf(
+    not importlib.util.find_spec("modelopt"),
```
Comment (narendasan) [inline tests/py/dynamo/models/test_models_export.py:207]: Wont we need blackwell as well?

File: py/torch_tensorrt/dynamo/_compiler.py (hunk 1 above)
Comment (peri044) [reply py/torch_tensorrt/dynamo/_compiler.py]: yes, the old message seems right. Is this required ?

File: py/torch_tensorrt/dynamo/conversion/_ConversionContext.py [hunk 13]
```diff
@@ -21,3 +23,6 @@ class ConversionContext:
     )
     requires_output_allocator: bool = False
     mapping: dict[str, np.array] = field(default_factory=dict)
+    weights_reference_holder: dict[str, Union[torch.Tensor, np.array]] = field(
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/conversion/_ConversionContext.py]: consider naming this to cpu_weights_reference_holder for more clarity

File: py/torch_tensorrt/dynamo/conversion/converter_utils.py [hunk 14]
```diff
@@ -361,12 +361,41 @@ def create_constant(
...
+                    type=trt.DataType.FP4,
+                    ptr=torch_value.data_ptr(),
+                    count=torch_value.numel() * 2,
+                )
+                constant = ctx.net.add_constant(
+                    shape,
+                    weights,
+                )
+                constant.name = name
+                ctx.weights_reference_holder[name + " FP4_CONSTANT"] = torch_value
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/conversion/converter_utils.py]: how do we distinguish a `uint8` constant (stored from FP4 workflow) vs a `uint8` constant manually inserted in the graph. (eg: the graph can have something like `torch.add(x, torch.tensor(2, dtype=torch.uint8))` ?

File: py/torch_tensorrt/dynamo/conversion/impl/dynamic_block_quantize.py [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+def _double_dequantize(
+    ctx: ConversionContext,
+    target: Target,
+    source_ir: Optional[SourceIR],
+    name: str,
+    quantized_data_in_fp4: TRTTensor,
+    quantized_scale_in_fp8: TRTTensor,
+    global_scale: torch.Tensor,
+    axis: int = -1,
+    output_type: trt.DataType = trt.DataType.FLOAT,
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/conversion/impl/dynamic_block_quantize.py:148]: minor : add a docstring and a little bit of explanation about the function wherever possible. Any references to this quantization (eg: TRT docs) will be useful

**Review by narendasan — COMMENTED** [review]
Add an end to end example for FP4 using modelopt and tensorrt

File: py/torch_tensorrt/dynamo/conversion/converter_utils.py (hunk 14 above)
Comment (narendasan) [reply py/torch_tensorrt/dynamo/conversion/converter_utils.py]: have an optional argument that sets the target type and in the case of uint8 without fp4 then throw error

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py (hunk 4 above)
Comment (narendasan) [reply py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py]: Just check it once

File: py/torch_tensorrt/dynamo/conversion/impl/dynamic_block_quantize.py (hunk 11 above)
Comment (lanluo-nvidia) [reply py/torch_tensorrt/dynamo/conversion/impl/dynamic_block_quantize.py:41]: below are all the information I can get from the aten ops,
I don't see any better way to know this is the weight or input
```
    @dynamo_tensorrt_converter(
        torch.ops.tensorrt.dynamic_block_quantize_op.default,
        supports_dynamic_shapes=True,
    )
    def aten_ops_dynamic_block_quantize_op(
        ctx: ConversionContext,
        target: Target,
        args: Tuple[Argument, ...],
        kwargs: Dict[str, Argument],
        name: str,
    ) -> Union[TRTTensor, Sequence[TRTTensor]]:
```

**Comment by lanluo-nvidia** [comment]
>Add an end to end example for FP4 using modelopt and tensorrt

There is a simple linear end to end FP4 example in this PR.
There is another FP4 flux example in a seperate PR: https://github.com/pytorch/TensorRT/pull/3537
Which based on Adrian's flux PR.
